### DISCUSSION TITLE
Future of NACK

### DISCUSSION THREAD

**Discussion by Jarema** [nats-io/nack#166] [Ideas] [answered]
This discussion is meant to align how the future of NACK should look like, and if there are NATS community members willing to contribute to make it happen.

# Current state of NACK

NACK, while working fine with the set of features it has, is also missing some important ones that are mentioned quite often:

1. No reconcile loop #95 
2. No support for KV and Object Store
3. Sometimes not clear defaults, or inheriting different defaults from the Server
4. Challenges in codebase with NACK working in a really declarative mode

# New opportunities

As nats 2.11 will get [Pedantic Mode for consumers and streams](https://github.com/nats-io/nats-server/issues/4959), and `jsm` will fully support it, it will become significantly easier to handle changes via any tooling that manages NATS resources in declarative way.

# Scope
The proposed scope for changes is as follows:

- add reconcile loop that checks streams and consumers state
- support streams, consumers, kv and object store
- add support for ownership via metadata
- cleanup / rewrite of the NACK codebase if it makes things much easier thanks to both Pedantic Mode and how we interact and test k8s layer.

Please share your feedback!

**Comment by mweibel** [comment]
That would be great. As it is now, I'm not sure I can run NACK on a production cluster because it doesn't support reconciliation. 

It would also be great to add admission webhooks for a few cases. 
For example:

```bash
# Create a stream.
$ kubectl apply -f - <<EOF
apiVersion: jetstream.nats.io/v1beta2
kind: Stream
metadata:
  name: mystream
spec:
  name: mystream
  subjects: ["orders.*"]
  storage: file
  maxAge: 1h
  replicas: 1
EOF

# create same stream with another metadata.name
$ kubectl apply -f - <<EOF
apiVersion: jetstream.nats.io/v1beta2
kind: Stream
metadata:
  name: mystream2
spec:
  name: mystream
  subjects: ["orders.*"]
  storage: file
  maxAge: 1h
  replicas: 1
EOF

# delete one of them
$ kubectl delete stream mystream2
```

When you do this, you'll see that stream `mystream` has been deleted even though another Stream is still referencing it. 

Using reconciliation the stream would get created again. However, in this case it might be more beneficial to deny creating another stream referencing the same stream name using an admission webhook. 

The same applies to consumers and probably other custom resources of NACK too. 

More of a general remark: I wonder if it makes sense to be able to specify the `spec.name` when there's already `metadata.name`. Same for consumers: does it make sense to be able to specify `spec.durableName` when `metadata.name` is already given by kubernetes?

**Comment by kejne** [comment, marked as answer]
We're using the NACK functionality quite a lot together with NAuth to create multiple accounts within the cluster.
The problem right now is the low community engagement of the project and that the project would benefit from more contributors to automate some of the setup.

I would be willing to contribute more in waves.

I believe this is a great area for non-Synadia contributors to also take part of the development as it's more geared towards knowledge of Kubernetes and operators rather than purely NATS.

However - should discussions be here, or should these also be moved to Slack? Although the 90-days visibility of Slack *sucks*...

**Comment by bobo** [comment]
Hi!
Is this ticket outdated?
From my understanding adding the --control-loop parameter to nack now enabled a reconcile loop.
Although that is flagged as experimental in the readme, it atleast exists and seems to work well.
That also adds support for kv and objectstore to my understanding.

Or is there sitll some objects that wont be reconciled?

This ticket almost stopped me from using nack...

**Reply by kejne** [reply]
@bobo Yes, or most of all the subject might not match the content.
This is not an issue, but rather only a discussion.

We're using `--control-loop` as well and it works as intended. Think that this should be considered the default and the old variant to be deprecated.

@Jarema - maybe you can mark this as "answered"? And maybe also move discussions to NATS slack since there is low activity here and has outdated threads.

**Comment by Jarema** [comment]
Yes, it is implemented. It will remain experimental for some time, until broad adoption helps resolve any potential edge cases, however this is a recommended way to run NACK.
